traffic_ops/traffic_ops_golang/cfgdiffs.go

```diff
@@ -0,0 +1,303 @@
+
+type CfgFileDiffs struct {
+	FileName         string   `json:"fileName"`
+	DBLinesMissing   []string `json:"dbLinesMissing"`
+	DiskLinesMissing []string `json:"diskLinesMissing"`
+	ReportTimestamp  string   `json:"timestamp"`
+}
+
+type CfgFileDiffsResponse struct {
+	Response []CfgFileDiffs `json:"response"`
+}
+
+type ServerExistsMethod func(db *sqlx.DB, hostname string) (bool, error)
+type UpdateCfgDiffsMethod func(db *sqlx.DB, hostname string, diffs CfgFileDiffs) (bool, error)
+type InsertCfgDiffsMethod func(db *sqlx.DB, hostname string, diffs CfgFileDiffs) error
+type GetCfgDiffsMethod func(db *sqlx.DB, hostName string) ([]CfgFileDiffs, error)
+
+func getCfgDiffsHandler(db *sqlx.DB) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		handleErr := func(err error, status int) {
+			log.Errorf("%v %v\n", r.RemoteAddr, err)
+			w.WriteHeader(status)
+			fmt.Fprintf(w, http.StatusText(status))
+		}
+		ctx := r.Context()
+		pathParams, err := getPathParams(ctx)
+		if err != nil {
+			handleErr(err, http.StatusInternalServerError)
+			return
+		}
+
+		hostName := pathParams["host-name"]
+
+		resp, err := getCfgDiffsJson(hostName, db, getCfgDiffs)
+		if err != nil {
+			handleErr(err, http.StatusInternalServerError)
+			return
+		}
+
+		// if the response has a length of zero, no results were found for that server
+		if len(resp.Response) == 0 {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+
+		respBts, err := json.Marshal(resp)
+		if err != nil {
+			handleErr(err, http.StatusInternalServerError)
+			return
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		fmt.Fprintf(w, "%s", respBts)
+	}
+}
+
+func putCfgDiffsHandler(db *sqlx.DB) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		handleErr := func(err error, status int) {
+			log.Errorf("%v %v\n", r.RemoteAddr, err)
+			w.WriteHeader(status)
+			fmt.Fprintf(w, http.StatusText(status))
+		}
+		ctx := r.Context()
+		pathParams, err := getPathParams(ctx)
+		if err != nil {
+			handleErr(err, http.StatusInternalServerError)
+			return
+		}
+
+		hostName := pathParams["host-name"]
+		configName := pathParams["cfg-file-name"]
+
+		decoder := json.NewDecoder(r.Body)
+		var diffs CfgFileDiffs
+		err = decoder.Decode(&diffs)
+		if err != nil {
+			handleErr(err, http.StatusBadRequest)
+			return
+		}
+
+		defer r.Body.Close()
+
+		diffs.FileName = configName
+
+		result, err := putCfgDiffs(db, hostName, diffs, serverExists, updateCfgDiffs, insertCfgDiffs)
+		if err != nil {
+			handleErr(err, http.StatusInternalServerError)
+			return
+		}
+
+		// Not found (invalid hostname)
+		if result == 0 { // This keeps happening
+			w.WriteHeader(404)
+			return
+		}
+		// Created (newly added)
+		if result == 1 {
+			w.WriteHeader(201)
+			return
+		}
+		// Updated (already existed)
+		if result == 2 {
+			w.WriteHeader(202)
+			return
+		}
+	}
+}
+
+func serverExists(db *sqlx.DB, hostName string) (bool, error) {
```

Comment:
shouldn't this also include `domainName` to make sure it's unique?

Reply:
Yea, I didn't know whether `host_name`s were unique or not.  Should the `host_name` and `domain_name` be sufficient?